src/documents/templates/DeploymentTemplateDoc.ts

```diff
@@ -477,10 +477,115 @@ export class DeploymentTemplateDoc extends DeploymentDocument {
+                    let span: Span.Span = new Span.Span(startIndex, endIndex - startIndex);
+                    const selectedText = this.getDocumentTextWithSquareBrackets(span);
+                    if (this.isParameterOrVariableReference(selectedText)) {
+                        return actions;
+                    }
+                    if (pc.jsonValue && jsonToken.value && jsonToken.value.span === pc.jsonValue.span && selectedText && this.equalsWithSqareBrackets(pc.jsonValue.asStringValue?.unquotedValue, selectedText)) {
+                        shouldAddExtractActions = true;
+                    } else {
+                        if (this.isExpression(value.quotedValue)) {
+                            shouldAddExtractActions = this.isValidExpression(this.getDocumentTextWithSurroundingCharacters(span, "'", "'"));
+                        }
+                    }
+                } else {
+                    if (this.isSimpleText(value.quotedValue) && pc.jsonValue && jsonToken.value && jsonToken.value.span === pc.jsonValue.span) {
+                        shouldAddExtractActions = true;
+                    }
+                }
+            }
+        }
+        if (shouldAddExtractActions) {
+            actions.push(this.createExtractCommand('Extract Parameter...', 'extractParameter'));
+            actions.push(this.createExtractCommand('Extract Variable...', 'extractVariable'));
+        }
         return actions;
     }
 
+    private getDocumentTextWithSquareBrackets(span: Span.Span): string {
+        let text = this.getDocumentText(span);
+        if (text.startsWith("[") && text.endsWith("]")) {
+            return text;
+        }
+        let extendedSpan = span.extendLeft(1).extendRight(1);
+        let extendedText = this.getDocumentText(extendedSpan);
+        if (extendedText.startsWith("[") && extendedText.endsWith("]")) {
+            return extendedText;
+        }
+        return text;
+    }
+
+    private getDocumentTextWithSurroundingCharacters(span: Span.Span, start: string, end: string): string {
+        let text = this.getDocumentText(span);
+        if (text.startsWith(start) && text.endsWith(end)) {
+            return text;
+        }
+        let extendedSpan = span.extendLeft(1).extendRight(1);
+        let extendedText = this.getDocumentText(extendedSpan);
+        if (extendedText.startsWith(start) && extendedText.endsWith(end)) {
+            return extendedText;
+        }
+        return text;
+    }
+
+    private isParameterOrVariableReference(text: string): boolean {
+        const regEx = /^"?\[?(parameters|variables)\('.+'\)]?"?/gi;
+        return regEx.test(text);
+    }
+
+    private isSimpleText(text: string): boolean {
+        if (text.startsWith("\"[") || text.endsWith("]\"")) {
+            return false;
+        }
+        return text.startsWith("\"") && text.endsWith("\"");
+    }
+
+    public isExpression(text: string): boolean {
```

Comment:
There should already be functions for stuff like this, such as isTleExpression

Reply:
Fixed,